src/test/java/com/waverleysoftware/BaseTest.java

```diff
@@ -1,37 +1,74 @@
 package com.waverleysoftware;
 
+import com.waverleysoftware.pageobjects.HomePage;
+import io.github.bonigarcia.wdm.WebDriverManager;
 import lombok.extern.slf4j.Slf4j;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.remote.BrowserType;
 import org.testng.ITestContext;
-import org.testng.annotations.AfterClass;
-import org.testng.annotations.BeforeClass;
-import org.testng.annotations.BeforeMethod;
-import org.testng.annotations.Test;
+import org.testng.annotations.*;
 
 import java.lang.reflect.Method;
 import java.time.LocalDate;
+import java.util.concurrent.TimeUnit;
+
+import static io.vavr.API.*;
 
 /**
  * A simple base class for processing tests.
  */
 @Slf4j
 public class BaseTest {
 
+    private static final ThreadLocal<WebDriver> WEB_DRIVER = new ThreadLocal<>();
+
     @BeforeClass
     public void setUp(final ITestContext context) {
-        log.info("{} - {}", "Starting process on parent level with {}",
+        log.info("{} - {}", "Starting process on parent level",
                 LocalDate.now(),
                 context.getCurrentXmlTest().getName());
     }
 
     @BeforeMethod
     public void setUp(final ITestContext context, final Method method) {
-        log.info("Method {}, with declared annotation {} and with group {}.",
-                method.getName(), String.valueOf(method.getDeclaredAnnotation(Test.class)),
-                context.getCurrentXmlTest().getParameter("browser"));
+        log.info("Method {}, with declared annotation {}",
+                method.getName(), String.valueOf(method.getDeclaredAnnotation(Test.class)));
+        WEB_DRIVER.set(Match(context.getCurrentXmlTest().getParameter("browser")).of(
+                Case($(BrowserType.FIREFOX), () -> {
+                    WebDriverManager.getInstance(FirefoxDriver.class).setup();
+                    return new FirefoxDriver();
+                }),
+                Case($(BrowserType.CHROME), () -> {
+                    WebDriverManager.getInstance(ChromeDriver.class).setup();
+                    return new ChromeDriver();
+                })
+        ));
+    }
+
+    @AfterMethod
+    public void tearDown(final ITestContext context) {
+        log.info("Ending process on parent method level");
+        if (getWebDriver() != null) {
+            getWebDriver().close();
+            getWebDriver().quit();
+        }
+        WEB_DRIVER.remove();
     }
 
     @AfterClass
     public void tearDown() {
-        log.info("Ending process on parent level");
+        log.info("Ending process on parent class level");
+    }
+
+    public static WebDriver getWebDriver() {
+        return WEB_DRIVER.get();
+    }
+
+    protected HomePage getHomePage() {
+        getWebDriver().navigate().to(System.getenv("URL"));
```

Comment:
you've put url into props file, but reading it from env var... it seems confusing